sdks/go/pkg/beam/core/runtime/exec/sdf_invokers_test.go

```diff
@@ -210,86 +229,154 @@ func TestInvokes(t *testing.T) {
 }
 
-// CreateInitialRestriction creates a restriction with the sum of the given
-// values.
-func (fn *KvSdf) CreateInitialRestriction(i int, j int) Restriction {
-	return Restriction{i + j}
+// CreateInitialRestriction creates a restriction with the given values and
+// with the appropriate flags to track that this was called.
+func (fn *VetKvSdf) CreateInitialRestriction(i, j int) *VetRestriction {
+	return &VetRestriction{ID: "KvSdf", Key: i, Val: j, CreateRest: true}
 }
 
-// SplitRestriction outputs two restrictions, the first containing the sum of i
-// and rest.Val, the second containing the sum of j and rest.Val.
-func (fn *KvSdf) SplitRestriction(i int, j int, rest Restriction) []Restriction {
-	return []Restriction{{rest.Val + i}, {rest.Val + j}}
+// SplitRestriction outputs two identical restrictions, each being a copy of the
+// initial one, but with the appropriate flags to track this was called. The
+// split restrictions add a suffix of the form ".#" to the ID.
+func (fn *VetKvSdf) SplitRestriction(i, j int, rest *VetRestriction) []*VetRestriction {
+	rest.SplitRest = true
+	rest1 := &VetRestriction{
+		ID:            rest.ID + ".1",
+		Key:           i,
+		Val:           j,
+		CreateRest:    rest.CreateRest,
+		SplitRest:     true,
+		RestSize:      rest.RestSize,
+		CreateTracker: rest.CreateTracker,
+		ProcessElm:    rest.ProcessElm,
+	}
+	rest2 := &VetRestriction{}
+	*rest2 = *rest1
+	rest2.ID = rest.ID + ".2"
+	return []*VetRestriction{rest1, rest2}
 }
 
-// RestrictionSize returns the sum of i, j, and rest.Val as a float64.
-func (fn *KvSdf) RestrictionSize(i int, j int, rest Restriction) float64 {
-	return (float64)(i + j + rest.Val)
+// RestrictionSize just returns the sum of i and j as the size, as well as
+// flipping appropriate flags on the restriction to track that this was called.
+func (fn *VetKvSdf) RestrictionSize(i, j int, rest *VetRestriction) float64 {
+	rest.Key = i
+	rest.Val = j
+	rest.RestSize = true
+	return (float64)(i + j)
 }
 
-// CreateTracker creates an RTracker containing the given restriction and a Val
-// of 2.
-func (fn *KvSdf) CreateTracker(rest Restriction) *RTracker {
-	return &RTracker{rest, 2}
+// CreateTracker creates an RTracker containing the given restriction and flips
+// the appropriate flags on the restriction to track that this was called.
+func (fn *VetKvSdf) CreateTracker(rest *VetRestriction) *VetRTracker {
+	rest.CreateTracker = true
+	return &VetRTracker{rest}
 }
 
-// ProcessElement emits two ints. The first is the first input (key) +
-// RTracker.Rest.Val. The second is the second input (value) + RTracker.Val.
-func (fn *KvSdf) ProcessElement(rt *RTracker, i1 int, i2 int, emit func(int, int)) {
-	emit(i1+rt.Rest.Val, i2+rt.Val)
+// ProcessElement emits a copy of the restriction in the restriction tracker it
+// received, with the appropriate flags flipped to track that this was called.
+func (fn *VetKvSdf) ProcessElement(rt *VetRTracker, i, j int, emit func(VetRestriction)) {
```

Comment:
It's a little awkward to have both Value instances of a type and Pointer instances of a type. Usually one is using one or the other. This makes it harder in to understand the test expectations, which if things need to change, could cause harder to diagnose errors.

Since you're using pointers to allow for the  "was this specific instance processed?" consider consolidating on the pointer types.

Reply:
Good point. I agree, consistency seems preferable here.